ddtrace/internal/flare.py

```diff
@@ -0,0 +1,184 @@
+import io
+import json
+import logging
+import os
+import secrets
+import shutil
+import tarfile
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Tuple
+
+from ddtrace import config
+from ddtrace._logger import _add_file_handler
+from ddtrace.internal.logger import _get_handler
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.utils.http import get_connection
+
+
+TRACER_FLARE_DIRECTORY = "tracer_flare"
+TRACER_FLARE_TAR = f"{TRACER_FLARE_DIRECTORY}.tar"
+TRACER_FLARE_ENDPOINT = "/tracer_flare/v1"
+TRACER_FLARE_FILE_HANDLER_NAME = "tracer_flare_file_handler"
+TRACER_FLARE_LOCK = "tracer_flare.lock"
+DEFAULT_TIMEOUT_SECONDS = 5
+
+log = get_logger(__name__)
+
+
+class Flare:
+    def __init__(self, timeout_sec: int = DEFAULT_TIMEOUT_SECONDS):
+        self.original_log_level = 0
+        self.timeout = timeout_sec
+
+    def _prepare(self, configs: List[dict]):
+        """
+        Update configurations to start sending tracer logs to a file
+        to be sent in a flare later.
+        """
+        os.makedirs(TRACER_FLARE_DIRECTORY, exist_ok=True)
+        for c in configs:
+            # AGENT_CONFIG is currently being used for multiple purposes
+            # We only want to prepare for a tracer flare if the config name
+            # starts with 'flare-log-level'
+            if not c.get("name", "").startswith("flare-log-level"):
+                return
+
+            # Validate the flare log level
+            flare_log_level = c.get("config", {}).get("log_level").upper()
+            flare_log_level_int = logging.getLevelName(flare_log_level)
+            if type(flare_log_level_int) != int:
+                raise TypeError("Invalid log level provided: %s", flare_log_level_int)
+
+            ddlogger = get_logger("ddtrace")
+            pid = os.getpid()
+            flare_file_path = f"{TRACER_FLARE_DIRECTORY}/tracer_python_{pid}.log"
+            self.original_log_level = ddlogger.level
+
+            # Set the logger level to the more verbose between original and flare
+            # We do this valid_original_level check because if the log level is NOTSET, the value is 0
+            # which is the minimum value. In this case, we just want to use the flare level, but still
+            # retain the original state as NOTSET/0
+            valid_original_level = 100 if self.original_log_level == 0 else self.original_log_level
+            logger_level = min(valid_original_level, flare_log_level_int)
+            ddlogger.setLevel(logger_level)
+            _add_file_handler(ddlogger, flare_file_path, flare_log_level, TRACER_FLARE_FILE_HANDLER_NAME)
```

Comment:
if you have `_add_file_handler` return the instance of the handler, it would make removing later easier.

```python
self.log_handler = _add_file_handler(...)

...


ddlogger.removeHandler(self.log_handler)
```

you also wouldn't need `_get_handler` either ?

Reply:
Ooh, nice. I'll do this instead.

Reply:
Actually, it's kinda handy for tests, so I'll just move it to the test file 😅